pilot/pkg/security/authz/builder/extauthz.go

```diff
@@ -0,0 +1,315 @@
+		}
+	}
+	if errs != nil {
+		return nil, errs
+	}
+
+	return generateHTTPConfig(hostname, cluster, status, config), nil
+}
+
+func buildExtAuthzGRPC(config *meshconfig.MeshConfig_ExtensionProvider_EnvoyExternalAuthorizationGrpcProvider) (*builtExtAuthz, error) {
+	var errs error
+	port, err := parsePort(config.Port)
+	if err != nil {
+		errs = multierror.Append(errs, err)
+	}
+	_, cluster, err := parseService(config.Service, port)
+	if err != nil {
+		errs = multierror.Append(errs, err)
+	}
+	status, err := parseStatusOnError(config.StatusOnError)
+	if err != nil {
+		errs = multierror.Append(errs, err)
+	}
+	if errs != nil {
+		return nil, errs
+	}
+
+	return generateGRPCConfig(cluster, config.FailOpen, status), nil
+}
+
+func parseService(service string, port int) (hostname string, cluster string, err error) {
+	if service == "" {
+		return "", "", fmt.Errorf("service must not be empty")
+	}
+	parts := strings.Split(service, "/")
+	if len(parts) > 2 {
+		return "", "", fmt.Errorf("service not in format [<namespace>/]<service>, found: %s", service)
+	}
+	var name, namespace string
+	if len(parts) == 2 {
+		namespace, name = parts[0], parts[1]
+	} else {
+		// TODO(yangminzhu): Fix to use the namespace of MeshConfig.
+		namespace, name = "istio-system", parts[0]
+	}
+
+	// TODO(yangminzhu): This is temporary and still under development.
+	// The remaining work is to properly get the corresponding cluster name for the k8s service and ServiceEntry.
+	hostname = fmt.Sprintf("%s.%s.svc.cluster.local", name, namespace)
+	cluster = model.BuildSubsetKey(model.TrafficDirectionOutbound, "", host.Name(hostname), port)
+	return hostname, cluster, nil
+}
+
+func parsePort(port uint32) (int, error) {
+	if 1 <= port && port <= 65535 {
+		return int(port), nil
+	}
+	return 0, fmt.Errorf("port must be in the range [1, 65535], found: %d", port)
+}
+
+func parseStatusOnError(status string) (*envoytypev3.HttpStatus, error) {
+	if status == "" {
+		return nil, nil
+	}
+	code, err := strconv.ParseInt(status, 10, 32)
+	if err != nil {
+		return nil, fmt.Errorf("invalid statusOnError %s: %v", status, err)
+	}
+	if _, found := envoytypev3.StatusCode_name[int32(code)]; !found {
+		return nil, fmt.Errorf("unsupported statusOnError %s, supported values: %v", status, envoytypev3.StatusCode_name)
+	}
+	return &envoytypev3.HttpStatus{Code: envoytypev3.StatusCode(code)}, nil
+}
+
+func generateHTTPConfig(hostname, cluster string, status *envoytypev3.HttpStatus,
+	config *meshconfig.MeshConfig_ExtensionProvider_EnvoyExternalAuthorizationHttpProvider) *builtExtAuthz {
+	service := &extauthzhttp.HttpService{
+		PathPrefix: config.PathPrefix,
+		ServerUri: &envoy_config_core_v3.HttpUri{
+			// Timeout is required, use a large value as a placeholder, the real timeout is controlled in DestinationRule.
+			Timeout: &duration.Duration{Seconds: 600},
+			// Uri is required but actually not used in the ext_authz filter.
```

Comment:
Sounds like ext authz filter's wrong constraint as well, as above?

(no AI, just confirm)

Reply:
ext_authz filter is using referring to the generic HttpUri API, not only validation makes sense in the ext_authz context.

Reply:
because envoy is validating as HTTP URI typed, i see.

"s/required/required for Envoy's validation/g" optional